python_modules/libraries/dagster-components/dagster_components/core/schema/objects.py

```diff
@@ -15,15 +16,26 @@
 
 from dagster_components.core.schema.base import ResolvableModel
 from dagster_components.core.schema.metadata import ResolvableFieldInfo
-from dagster_components.core.schema.resolver import TemplatedValueResolver
+from dagster_components.core.schema.resolver import ResolveContext
 
 
-class OpSpecModel(ResolvableModel["OpSpecModel"]):
+class OpSpecModel(ResolvableModel):
     name: Optional[str] = None
     tags: Optional[dict[str, str]] = None
 
-    def resolve(self, resolver: TemplatedValueResolver) -> "OpSpecModel":
-        return OpSpecModel(**resolver.resolve_obj(self.model_dump(exclude_unset=True)))
+    def resolve(self, context: ResolveContext) -> "OpSpecModel":
+        return self.resolve_as(OpSpecModel, context)
+
+
+class AssetDepModel(ResolvableModel):
+    asset: str
+    partition_mapping: Optional[str] = None
+
+    def resolve_asset(self, context: ResolveContext) -> AssetKey:
+        return AssetKey.from_coercible(context.resolve_value(self.asset))
+
+    def resolve(self, context: ResolveContext) -> AssetDep:
+        return self.resolve_as(AssetDep, context)
```

Comment:
is there no way to make a default implementation of this in the base class?

Reply:
Updated to do this with some generic shenanigans.

There will be plenty of cases where this is not possible (i.e. you're resolving to something other than a simple type), but in those cases you just have to override the `resolve` method, which isn't so bad